# Notebook 1: G2P and IPA
## What this notebook covers
- How text gets converted to phonemes (G2P = Grapheme to Phoneme)
- What IPA is and why it matters
- Why names break standard pronunciation systems
- The exact problem NameCoach Euphonia is solving

Built as a practical study exercise before the take-home.

In [1]:
# Libraries already installed via pip in terminal
# espeak-ng installed separately via the Windows installer
# This cell just verifies imports work

import nltk
import jiwer
from phonemizer import phonemize

print("All imports successful.")


A module that was compiled using NumPy 1.x cannot be run in
NumPy 2.2.6 as it may crash. To support both 1.x and 2.x
versions of NumPy, modules must be compiled with NumPy 2.0.
Some module may need to rebuild instead e.g. with 'pybind11>=2.12'.

If you are a user of the module, the easiest solution will be to
downgrade to 'numpy<2' or try to upgrade the affected module.
We expect that some modules will need time to support NumPy 2.

Traceback (most recent call last):  File "<frozen runpy>", line 198, in _run_module_as_main
  File "<frozen runpy>", line 88, in _run_code
  File "C:\Users\garvd\anaconda3\Lib\site-packages\ipykernel_launcher.py", line 17, in <module>
    app.launch_new_instance()
  File "C:\Users\garvd\anaconda3\Lib\site-packages\traitlets\config\application.py", line 1075, in launch_instance
    app.start()
  File "C:\Users\garvd\anaconda3\Lib\site-packages\ipykernel\kernelapp.py", line 701, in start
    self.io_loop.start()
  File "C:\Users\garvd\anaconda3\Lib\site-pack

ImportError: 
A module that was compiled using NumPy 1.x cannot be run in
NumPy 2.2.6 as it may crash. To support both 1.x and 2.x
versions of NumPy, modules must be compiled with NumPy 2.0.
Some module may need to rebuild instead e.g. with 'pybind11>=2.12'.

If you are a user of the module, the easiest solution will be to
downgrade to 'numpy<2' or try to upgrade the affected module.
We expect that some modules will need time to support NumPy 2.




A module that was compiled using NumPy 1.x cannot be run in
NumPy 2.2.6 as it may crash. To support both 1.x and 2.x
versions of NumPy, modules must be compiled with NumPy 2.0.
Some module may need to rebuild instead e.g. with 'pybind11>=2.12'.

If you are a user of the module, the easiest solution will be to
downgrade to 'numpy<2' or try to upgrade the affected module.
We expect that some modules will need time to support NumPy 2.

Traceback (most recent call last):  File "<frozen runpy>", line 198, in _run_module_as_main
  File "<frozen runpy>", line 88, in _run_code
  File "C:\Users\garvd\anaconda3\Lib\site-packages\ipykernel_launcher.py", line 17, in <module>
    app.launch_new_instance()
  File "C:\Users\garvd\anaconda3\Lib\site-packages\traitlets\config\application.py", line 1075, in launch_instance
    app.start()
  File "C:\Users\garvd\anaconda3\Lib\site-packages\ipykernel\kernelapp.py", line 701, in start
    self.io_loop.start()
  File "C:\Users\garvd\anaconda3\Lib\site-pack

ImportError: 
A module that was compiled using NumPy 1.x cannot be run in
NumPy 2.2.6 as it may crash. To support both 1.x and 2.x
versions of NumPy, modules must be compiled with NumPy 2.0.
Some module may need to rebuild instead e.g. with 'pybind11>=2.12'.

If you are a user of the module, the easiest solution will be to
downgrade to 'numpy<2' or try to upgrade the affected module.
We expect that some modules will need time to support NumPy 2.




A module that was compiled using NumPy 1.x cannot be run in
NumPy 2.2.6 as it may crash. To support both 1.x and 2.x
versions of NumPy, modules must be compiled with NumPy 2.0.
Some module may need to rebuild instead e.g. with 'pybind11>=2.12'.

If you are a user of the module, the easiest solution will be to
downgrade to 'numpy<2' or try to upgrade the affected module.
We expect that some modules will need time to support NumPy 2.

Traceback (most recent call last):  File "<frozen runpy>", line 198, in _run_module_as_main
  File "<frozen runpy>", line 88, in _run_code
  File "C:\Users\garvd\anaconda3\Lib\site-packages\ipykernel_launcher.py", line 17, in <module>
    app.launch_new_instance()
  File "C:\Users\garvd\anaconda3\Lib\site-packages\traitlets\config\application.py", line 1075, in launch_instance
    app.start()
  File "C:\Users\garvd\anaconda3\Lib\site-packages\ipykernel\kernelapp.py", line 701, in start
    self.io_loop.start()
  File "C:\Users\garvd\anaconda3\Lib\site-pack

ImportError: 
A module that was compiled using NumPy 1.x cannot be run in
NumPy 2.2.6 as it may crash. To support both 1.x and 2.x
versions of NumPy, modules must be compiled with NumPy 2.0.
Some module may need to rebuild instead e.g. with 'pybind11>=2.12'.

If you are a user of the module, the easiest solution will be to
downgrade to 'numpy<2' or try to upgrade the affected module.
We expect that some modules will need time to support NumPy 2.




A module that was compiled using NumPy 1.x cannot be run in
NumPy 2.2.6 as it may crash. To support both 1.x and 2.x
versions of NumPy, modules must be compiled with NumPy 2.0.
Some module may need to rebuild instead e.g. with 'pybind11>=2.12'.

If you are a user of the module, the easiest solution will be to
downgrade to 'numpy<2' or try to upgrade the affected module.
We expect that some modules will need time to support NumPy 2.

Traceback (most recent call last):  File "<frozen runpy>", line 198, in _run_module_as_main
  File "<frozen runpy>", line 88, in _run_code
  File "C:\Users\garvd\anaconda3\Lib\site-packages\ipykernel_launcher.py", line 17, in <module>
    app.launch_new_instance()
  File "C:\Users\garvd\anaconda3\Lib\site-packages\traitlets\config\application.py", line 1075, in launch_instance
    app.start()
  File "C:\Users\garvd\anaconda3\Lib\site-packages\ipykernel\kernelapp.py", line 701, in start
    self.io_loop.start()
  File "C:\Users\garvd\anaconda3\Lib\site-pack

ImportError: 
A module that was compiled using NumPy 1.x cannot be run in
NumPy 2.2.6 as it may crash. To support both 1.x and 2.x
versions of NumPy, modules must be compiled with NumPy 2.0.
Some module may need to rebuild instead e.g. with 'pybind11>=2.12'.

If you are a user of the module, the easiest solution will be to
downgrade to 'numpy<2' or try to upgrade the affected module.
We expect that some modules will need time to support NumPy 2.



All imports successful.


In [2]:
import nltk

# Download CMUdict if not already downloaded
nltk.download('cmudict', quiet=True)

# Load it into memory
from nltk.corpus import cmudict
cmu = cmudict.dict()

# Sanity check
print(f"CMUdict loaded successfully.")
print(f"Total words in dictionary: {len(cmu)}")
print()
print(f"Example 1 — 'hello':  {cmu['hello']}")
print(f"Example 2 — 'world':  {cmu['world']}")
print(f"Example 3 — 'python': {cmu['python']}")

CMUdict loaded successfully.
Total words in dictionary: 123455

Example 1 — 'hello':  [['HH', 'AH0', 'L', 'OW1'], ['HH', 'EH0', 'L', 'OW1']]
Example 2 — 'world':  [['W', 'ER1', 'L', 'D']]
Example 3 — 'python': [['P', 'AY1', 'TH', 'AA0', 'N']]


In [3]:
def lookup_cmu(word):
    """
    Look up a word's phonemes in CMUdict.
    
    Returns the first (most common) pronunciation if found.
    Returns None if the word is not in the dictionary.
    
    Why .lower()? CMUdict stores everything in lowercase.
    Why .strip()? Removes accidental spaces around the word.
    Why result[0]? Returns just the first pronunciation.
    """
    word = word.lower().strip()
    result = cmu.get(word)
    
    if result:
        return result[0]   # First pronunciation only
    else:
        return None        # Word not found

# Test on common English words first
test_words = ["hello", "world", "pronunciation", "python", "engineer"]

print("Common English words:\n")
for word in test_words:
    phonemes = lookup_cmu(word)
    # The :20 pads the word to 20 characters so columns line up neatly
    print(f"  {word:20} → {phonemes}")

Common English words:

  hello                → ['HH', 'AH0', 'L', 'OW1']
  world                → ['W', 'ER1', 'L', 'D']
  pronunciation        → ['P', 'R', 'OW0', 'N', 'AH2', 'N', 'S', 'IY0', 'EY1', 'SH', 'AH0', 'N']
  python               → ['P', 'AY1', 'TH', 'AA0', 'N']
  engineer             → ['EH1', 'N', 'JH', 'AH0', 'N', 'IH1', 'R']


In [4]:
# These are real names from different cultural origins
# that appear constantly in healthcare, education, and enterprise contexts
# and that voice agents regularly mispronounce

names = [
    # Irish names
    "Saoirse",      # Pronounced: SUR-sha
    "Siobhan",      # Pronounced: Shih-VAWN  
    "Aoife",        # Pronounced: EE-fah
    "Caoimhe",      # Pronounced: KEE-vah

    # South Asian names
    "Priya",        # Pronounced: PREE-yah
    "Aarav",        # Pronounced: AH-rahv
    "Ananya",       # Pronounced: Ah-NUN-yah
    "Arjun",        # Pronounced: AR-jun

    # East/Southeast Asian names
    "Nguyen",       # Pronounced: Win (Vietnamese)
    "Linh",         # Pronounced: Lin
    "Yusuf",        # Arabic origin
    "Fatima",       # Arabic origin
]

print("CMUdict lookup results for names from different origins:")
print("=" * 60)

found = 0
not_found = 0

for name in names:
    result = lookup_cmu(name)
    if result:
        print(f"  FOUND    {name:15} → {result}")
        found += 1
    else:
        print(f"  MISSING  {name:15} → not in dictionary")
        not_found += 1

print("=" * 60)
print(f"\nFound:   {found}/{len(names)} names")
print(f"Missing: {not_found}/{len(names)} names")
print(f"\nEvery MISSING name will be mispronounced by a TTS system.")
print(f"This is the exact problem Euphonia exists to solve.")

CMUdict lookup results for names from different origins:
  MISSING  Saoirse         → not in dictionary
  FOUND    Siobhan         → ['SH', 'AW1', 'B', 'AA2', 'N']
  MISSING  Aoife           → not in dictionary
  MISSING  Caoimhe         → not in dictionary
  MISSING  Priya           → not in dictionary
  MISSING  Aarav           → not in dictionary
  MISSING  Ananya          → not in dictionary
  MISSING  Arjun           → not in dictionary
  FOUND    Nguyen          → ['N', 'UW0', 'Y', 'EH1', 'N']
  FOUND    Linh            → ['L', 'IH1', 'N']
  MISSING  Yusuf           → not in dictionary
  FOUND    Fatima          → ['F', 'AE1', 'TH', 'IH0', 'M', 'AH0']

Found:   4/12 names
Missing: 8/12 names

Every MISSING name will be mispronounced by a TTS system.
This is the exact problem Euphonia exists to solve.


In [5]:
from phonemizer import phonemize

def phonemize_word(word):
    """
    Use phonemizer with espeak-ng backend to convert
    a word to IPA (International Phonetic Alphabet).
    
    IPA is the global standard for representing pronunciation.
    Unlike ARPAbet (CMUdict), IPA works across all languages.
    
    backend="espeak" uses espeak-ng under the hood.
    language="en-us" tells it to use American English rules.
    strip=True removes extra whitespace from output.
    """
    result = phonemize(
        word,
        backend="espeak",
        language="en-us",
        strip=True
    )
    return result

# Test on common words first so you can see what IPA looks like
print("IPA output for common words:")
print("-" * 40)
common = ["hello", "world", "engineer", "pronunciation"]
for word in common:
    ipa = phonemize_word(word)
    print(f"  {word:20} → {ipa}")

print()
print("Now the names CMUdict missed:")
print("-" * 40)

missing_names = ["Saoirse", "Aoife", "Caoimhe", "Priya", 
                 "Aarav", "Ananya", "Arjun", "Yusuf"]

for name in missing_names:
    ipa = phonemize_word(name)
    print(f"  {name:15} → {ipa}")

IPA output for common words:
----------------------------------------
  hello                → həloʊ
  world                → wɜːld
  engineer             → ɛndʒɪnɪɹ
  pronunciation        → pɹənʌnsɪeɪʃən

Now the names CMUdict missed:
----------------------------------------
  Saoirse         → sɜːʃə
  Aoife           → iːfə
  Caoimhe         → kiːvə
  Priya           → pɹɪjə
  Aarav           → ɑːɹɹæv
  Ananya          → ænænjə
  Arjun           → ɑːɹdʒʌn
  Yusuf           → juːsʌf


In [6]:
def g2p(word):
    """
    Combined G2P function.
    
    Strategy:
    1. Try CMUdict first. It's fast, zero compute, and accurate
       for the 123k words it knows.
    2. If CMUdict misses, fall back to phonemizer/espeak.
       It handles unknown words including names, but with
       lower accuracy for non-English origins.
    
    Returns a dict with the phonemes, which source was used,
    and the notation type so we know what we're looking at.
    """
    word_clean = word.lower().strip()
    
    # Try CMUdict first
    cmu_result = cmu.get(word_clean)
    if cmu_result:
        return {
            "word": word,
            "phonemes": cmu_result[0],
            "source": "CMUdict",
            "notation": "ARPAbet",
            "found": True
        }
    
    # Fall back to phonemizer
    ipa_result = phonemize(
        word,
        backend="espeak",
        language="en-us",
        strip=True
    )
    return {
        "word": word,
        "phonemes": ipa_result,
        "source": "phonemizer/espeak",
        "notation": "IPA",
        "found": False   # Not in CMUdict, so accuracy lower
    }

# Test it across a mix of common words and names
test_set = [
    "hello",
    "engineer", 
    "Saoirse",
    "Priya",
    "pronunciation",
    "Arjun",
    "world",
    "Nguyen"
]

print("Combined G2P results:")
print("=" * 65)
print(f"  {'Word':<15} {'Source':<20} {'Phonemes'}")
print("=" * 65)

for word in test_set:
    result = g2p(word)
    confidence = "HIGH" if result["found"] else "LOW"
    print(f"  {word:<15} {result['source']:<20} {result['phonemes']}")
    print(f"  {'':15} Confidence: {confidence}, Notation: {result['notation']}")
    print()

Combined G2P results:
  Word            Source               Phonemes
  hello           CMUdict              ['HH', 'AH0', 'L', 'OW1']
                  Confidence: HIGH, Notation: ARPAbet

  engineer        CMUdict              ['EH1', 'N', 'JH', 'AH0', 'N', 'IH1', 'R']
                  Confidence: HIGH, Notation: ARPAbet

  Saoirse         phonemizer/espeak    sɜːʃə
                  Confidence: LOW, Notation: IPA

  Priya           phonemizer/espeak    pɹɪjə
                  Confidence: LOW, Notation: IPA

  pronunciation   CMUdict              ['P', 'R', 'OW0', 'N', 'AH2', 'N', 'S', 'IY0', 'EY1', 'SH', 'AH0', 'N']
                  Confidence: HIGH, Notation: ARPAbet

  Arjun           phonemizer/espeak    ɑːɹdʒʌn
                  Confidence: LOW, Notation: IPA

  world           CMUdict              ['W', 'ER1', 'L', 'D']
                  Confidence: HIGH, Notation: ARPAbet

  Nguyen          CMUdict              ['N', 'UW0', 'Y', 'EH1', 'N']
                  Confidence: HIGH

In [8]:
def edit_distance(seq1, seq2):
    """
    Compute the minimum number of edits (insert, delete, substitute)
    needed to turn seq1 into seq2.
    
    This is the Wagner-Fischer algorithm, the same one used
    for spell-checking, DNA comparison, and speech evaluation.
    
    Think of it as: how many single-phoneme changes does it take
    to turn the reference pronunciation into what was spoken?
    """
    m, n = len(seq1), len(seq2)
    
    # Create a grid of size (m+1) x (n+1)
    # dp[i][j] = edit distance between seq1[:i] and seq2[:j]
    dp = [[0] * (n + 1) for _ in range(m + 1)]
    
    # Base cases:
    # Turning seq1[:i] into empty string = i deletions
    for i in range(m + 1):
        dp[i][0] = i
    # Turning empty string into seq2[:j] = j insertions
    for j in range(n + 1):
        dp[0][j] = j
    
    # Fill the grid
    for i in range(1, m + 1):
        for j in range(1, n + 1):
            if seq1[i-1] == seq2[j-1]:
                # Phonemes match, no edit needed
                dp[i][j] = dp[i-1][j-1]
            else:
                dp[i][j] = 1 + min(
                    dp[i-1][j],    # deletion
                    dp[i][j-1],    # insertion
                    dp[i-1][j-1]   # substitution
                )
    
    return dp[m][n]


def phoneme_error_rate(reference, hypothesis):
    """
    PER = edit_distance(reference, hypothesis) / len(reference)
    
    reference:  the correct phoneme sequence (what SHOULD be said)
    hypothesis: the observed phoneme sequence (what WAS said)
    
    PER = 0.0 means perfect pronunciation
    PER = 1.0 means every phoneme was wrong
    PER > 1.0 is possible if hypothesis is much longer than reference
    
    We convert strings to lists of characters for IPA,
    or use the list directly for ARPAbet.
    """
    # Handle both ARPAbet lists and IPA strings
    if isinstance(reference, str):
        ref = list(reference)
    else:
        ref = reference
        
    if isinstance(hypothesis, str):
        hyp = list(hypothesis)
    else:
        hyp = hypothesis
    
    if len(ref) == 0:
        return 0.0
    
    distance = edit_distance(ref, hyp)
    per = distance / len(ref)
    return round(per, 3)


# Test 1: Perfect match
ref  = ['HH', 'AH0', 'L', 'OW1']
hyp1 = ['HH', 'AH0', 'L', 'OW1']
print(f"Perfect match:      PER = {phoneme_error_rate(ref, hyp1)}")

# Test 2: One phoneme wrong
hyp2 = ['HH', 'EH0', 'L', 'OW1']
print(f"One phoneme wrong:  PER = {phoneme_error_rate(ref, hyp2)}")

# Test 3: One phoneme missing
hyp3 = ['HH', 'L', 'OW1']
print(f"One phoneme missing: PER = {phoneme_error_rate(ref, hyp3)}")

# Test 4: Completely wrong
hyp4 = ['W', 'ER1', 'L', 'D']
print(f"Completely wrong:   PER = {phoneme_error_rate(ref, hyp4)}")

print()

# Test on IPA strings (for names)
print("IPA-based PER for Saoirse:")
correct_ipa   = "sɜːʃə"   # close to correct: SUR-sha
espeak_guess  = "sɜːʃə"   # what espeak gave us
wrong_attempt = "sɑːɪɹs"  # what a naive TTS might produce

print(f"  espeak vs correct: PER = {phoneme_error_rate(correct_ipa, espeak_guess)}")
print(f"  wrong vs correct:  PER = {phoneme_error_rate(correct_ipa, wrong_attempt)}")

Perfect match:      PER = 0.0
One phoneme wrong:  PER = 0.25
One phoneme missing: PER = 0.25
Completely wrong:   PER = 0.75

IPA-based PER for Saoirse:
  espeak vs correct: PER = 0.0
  wrong vs correct:  PER = 0.8


In [9]:
class PronunciationEvaluator:
    """
    End-to-end pronunciation evaluation pipeline.
    
    Given a word and a "spoken" phoneme sequence,
    tells you how accurately the word was pronounced.
    
    In a real system, the spoken phonemes would come from
    running audio through a phoneme recogniser like allosaurus
    or wav2vec2. Here we simulate that with manual input
    so you can understand the evaluation logic first.
    """
    
    def __init__(self):
        # Load CMUdict once at init, reuse for all evaluations
        import nltk
        nltk.download('cmudict', quiet=True)
        from nltk.corpus import cmudict
        self.cmu = cmudict.dict()
        print("PronunciationEvaluator ready.")
    
    def get_reference(self, word):
        """Get the reference (correct) phonemes for a word."""
        result = self.cmu.get(word.lower().strip())
        if result:
            return {
                "phonemes": result[0],
                "source": "CMUdict",
                "notation": "ARPAbet",
                "confidence": "HIGH"
            }
        
        ipa = phonemize(
            word, 
            backend="espeak", 
            language="en-us", 
            strip=True
        )
        return {
            "phonemes": list(ipa),
            "source": "phonemizer",
            "notation": "IPA",
            "confidence": "LOW"
        }
    
    def evaluate(self, word, spoken_phonemes):
        """
        Evaluate how accurately a word was pronounced.
        
        word:            the text of the word
        spoken_phonemes: what was actually said, as a phoneme list
        
        Returns a full evaluation report.
        """
        reference = self.get_reference(word)
        ref_phonemes = reference["phonemes"]
        
        per = phoneme_error_rate(ref_phonemes, spoken_phonemes)
        
        # Score from 0 to 100, easier to read than 0.0 to 1.0
        score = round((1 - min(per, 1.0)) * 100)
        
        # Grade the result
        if per == 0.0:
            grade = "PERFECT"
        elif per <= 0.2:
            grade = "GOOD"
        elif per <= 0.5:
            grade = "ACCEPTABLE"
        else:
            grade = "POOR"
        
        return {
            "word": word,
            "reference_phonemes": ref_phonemes,
            "spoken_phonemes": spoken_phonemes,
            "reference_source": reference["source"],
            "reference_confidence": reference["confidence"],
            "per_score": per,
            "accuracy_score": score,
            "grade": grade
        }
    
    def print_report(self, result):
        """Pretty print an evaluation result."""
        print(f"\nWord:           {result['word']}")
        print(f"Reference:      {result['reference_phonemes']}")
        print(f"  Source:       {result['reference_source']} "
              f"(confidence: {result['reference_confidence']})")
        print(f"Spoken:         {result['spoken_phonemes']}")
        print(f"PER:            {result['per_score']}")
        print(f"Accuracy:       {result['accuracy_score']}/100")
        print(f"Grade:          {result['grade']}")
        print("-" * 50)


# Create the evaluator
evaluator = PronunciationEvaluator()

print("\n" + "=" * 50)
print("SIMULATION: Testing pronunciation evaluation")
print("=" * 50)
print("(In production, spoken_phonemes would come")
print(" from running audio through a phoneme recogniser)")
print()

# Test 1: Perfect pronunciation of 'hello'
r1 = evaluator.evaluate(
    word="hello",
    spoken_phonemes=['HH', 'AH0', 'L', 'OW1']
)
evaluator.print_report(r1)

# Test 2: Slightly off pronunciation of 'hello'
r2 = evaluator.evaluate(
    word="hello",
    spoken_phonemes=['HH', 'EH0', 'L', 'OW1']
)
evaluator.print_report(r2)

# Test 3: Simulate someone saying 'Arjun' incorrectly
# Correct IPA from espeak: ɑːɹdʒʌn
# Wrong attempt: AR-jun with English R
arjun_ref = list(phonemize("Arjun", backend="espeak", 
                            language="en-us", strip=True))
arjun_wrong = list("ɑɹdʒʌn")  # shorter, missing ː

r3 = evaluator.evaluate(
    word="Arjun",
    spoken_phonemes=arjun_wrong
)
evaluator.print_report(r3)

# Test 4: Name with HIGH confidence source vs LOW confidence
r4 = evaluator.evaluate(
    word="Priya",
    spoken_phonemes=list("pɹɪjɑː")  # close but wrong vowel at end
)
evaluator.print_report(r4)

PronunciationEvaluator ready.

SIMULATION: Testing pronunciation evaluation
(In production, spoken_phonemes would come
 from running audio through a phoneme recogniser)


Word:           hello
Reference:      ['HH', 'AH0', 'L', 'OW1']
  Source:       CMUdict (confidence: HIGH)
Spoken:         ['HH', 'AH0', 'L', 'OW1']
PER:            0.0
Accuracy:       100/100
Grade:          PERFECT
--------------------------------------------------

Word:           hello
Reference:      ['HH', 'AH0', 'L', 'OW1']
  Source:       CMUdict (confidence: HIGH)
Spoken:         ['HH', 'EH0', 'L', 'OW1']
PER:            0.25
Accuracy:       75/100
Grade:          ACCEPTABLE
--------------------------------------------------

Word:           Arjun
Reference:      ['ɑ', 'ː', 'ɹ', 'd', 'ʒ', 'ʌ', 'n']
  Source:       phonemizer (confidence: LOW)
Spoken:         ['ɑ', 'ɹ', 'd', 'ʒ', 'ʌ', 'n']
PER:            0.143
Accuracy:       86/100
Grade:          GOOD
--------------------------------------------------

Word

## Summary and Key Findings

### What this notebook demonstrates

1. **The OOV Problem is severe**: CMUdict covers 123,455 English words
   but fails on 8/12 names from non-English origins (67% failure rate).

2. **"Found" does not equal "correct"**: Nguyen is in CMUdict but the
   entry is wrong. High confidence does not guarantee accuracy for
   names from non-English origins.

3. **phonemizer bridges the gap**: espeak-ng makes educated guesses
   for unknown words, getting roughly 3-4/8 names approximately right.
   But "approximately" is not good enough for a person's name.

4. **PER quantifies the gap**: A PER of 0.25 means one in four
   phonemes was wrong. For a 4-phoneme name, that is one sound.
   One wrong sound in a name can make it unrecognisable.

### Why NameCoach's dataset matters

No algorithm — CMUdict or phonemizer — reliably handles names from
non-English origins. The only robust solution is a human-verified
pronunciation dataset. That is what Euphonia is built on.

### What's next

- Notebook 2: Phoneme Error Rate in depth, edge cases, and using jiwer
- Notebook 3: Audio to phonemes with allosaurus
- Notebook 4: Full pipeline with real audio input

**Stack**: Python 3.12, nltk/CMUdict, phonemizer 3.4, espeak-ng 1.52